src/service/nickname.ts

```diff
@@ -0,0 +1,89 @@
+import {
+    Collection,
+    GuildMember,
+    OAuth2Guild,
+    PartialGuildMember,
+} from "discord.js";
+import { getUser } from "../wrapper/wrapper";
+import { config } from "../config";
+
+import { rankMap } from "../service/util/constants";
+import { UserRankRepository } from "../repository/UserRankRepository";
+import { UserRank } from "../entity/UserRank";
+import { UserData } from "../wrapper/models/user";
+import { Repository } from "typeorm";
+import { makeLogger } from "../logger";
+
+const logger = makeLogger(module);
+
+export async function processUser(
+    u: UserRank,
+    guilds: Collection<string, OAuth2Guild>
+) {
+    try {
+        const user = await getUser(config.apiKey, u.discord_id);
+        if ((guilds && user.name != u.user_name) || user.rank != u.rank) {
+            u.guild_ids.forEach(async (g) => {
+                const guild = await guilds.get(g)?.fetch();
+                if (guild) {
+                    const member = await guild.members.fetch(u.discord_id);
+                    processMember(member, user, u.id);
+                }
+            });
+        }
+    } catch (e) {
+        logger.debug(e);
+    }
+}
+
+/**
+ * @todo Add rank role
+ */
+async function processMember(member: GuildMember, user: UserData, id: string) {
+    if (member?.manageable) {
+        const rank = rankMap.get(user.rank) ?? "";
+        await member.edit({
+            nick: `${rank} ${user.name}`,
+        });
+        UserRankRepository.updateNameAndRank(id, user.rank, user.name);
+        logger.debug(
+            `Updated user with name ${user.name} and rank #${user.rank}`
+        );
+    }
+}
+
+export async function buildMember(
+    member: GuildMember,
+    repository: Repository<UserRank>
+) {
+    const userRank = await repository.findOne({
+        where: { discord_id: member.id },
+    });
+    const guild = member.guild.id;
+    if (userRank === null) {
+        const user = new UserRank();
```

Comment:
shouldn't this also fetch the name and rank from the api and fill it in? Seems to me like this should be tied to verification

Reply:
it will when members join a server (in the other file), but not when the bot is added to a server. imo its pointless to do it when the bot is added since it likely wont be able to edit anyones nicknames until the bot role is moved above the existing role.

Reply:
I mean, ultimately I think this feature should be an opt-in setting. That's easy to add onto this later on (I'm considering merging all the admin commands into one command group), but that would sidestep that specific scenario